packages/components/react/src/Panels/WorkspacePanel.tsx

```diff
@@ -71,157 +106,147 @@ export function WorkspacePanel({ tutorialStore, theme }: Props) {
     } else {
       tutorialStore.reset();
     }
-  }, [tutorialStore.ref]);
+  }
 
   useEffect(() => {
-    const lesson = tutorialStore.lesson!;
-
-    const unsubscribe = tutorialStore.lessonFullyLoaded.subscribe((loaded) => {
-      if (loaded && lesson.data.autoReload) {
-        /**
-         * @todo This causes some race with the preview where the iframe can show the "wrong" page.
-         * I think the reason is that when the ports are different then we render new frames which
-         * races against the reload which will internally reset the `src` attribute.
-         */
-        // previewRef.current?.reload();
-      }
-    });
-
     if (tutorialStore.hasSolution()) {
       setHelpAction('solve');
     } else {
       setHelpAction('reset');
     }
-
-    return () => unsubscribe();
   }, [tutorialStore.ref]);
 
-  useEffect(() => {
-    if (hideTerminalPanel) {
-      // force hide the terminal if we don't have any panels to show
-      hideTerminal();
+  return (
+    <Panel
+      id={hasEditor ? 'editor-opened' : 'editor-closed'}
+      defaultSize={hasEditor ? 50 : 0}
+      minSize={10}
+      maxSize={hasEditor ? 100 : 0}
+      collapsible={!hasEditor}
+      className="transition-theme bg-tk-elements-panel-backgroundColor text-tk-elements-panel-textColor"
+    >
+      <EditorPanel
+        id={tutorialStore.ref}
+        theme={theme}
+        showFileTree={tutorialStore.hasFileTree()}
+        editorDocument={currentDocument}
+        files={lesson.files[1]}
+        i18n={lesson.data.i18n as I18n}
+        hideRoot={lesson.data.hideRoot}
+        helpAction={helpAction}
+        onHelpClick={onHelpClick}
+        onFileSelect={(filePath) => tutorialStore.setSelectedFile(filePath)}
+        selectedFile={selectedFile}
+        onEditorScroll={(position) => tutorialStore.setCurrentDocumentScrollPosition(position)}
+        onEditorChange={(update) => tutorialStore.setCurrentDocumentContent(update.content)}
+      />
+    </Panel>
+  );
+}
 
-      terminalExpanded.current = false;
-    }
-  }, [hideTerminalPanel]);
+function Preview({
+  tutorialStore,
+  terminalPanelRef,
+  terminalExpanded,
+  hideTerminalPanel,
+  hasPreviews,
+  hasEditor,
+}: TerminalProps) {
+  const lesson = tutorialStore.lesson!;
 
-  const showTerminal = useCallback(() => {
+  const toggleTerminal = useCallback(() => {
     const { current: terminal } = terminalPanelRef;
 
     if (!terminal) {
       return;
     }
 
-    if (!terminalExpanded.current) {
-      terminalExpanded.current = true;
-      terminal.resize(DEFAULT_TERMINAL_SIZE);
+    if (terminalPanelRef.current?.isCollapsed()) {
+      if (!terminalExpanded.current) {
+        terminalExpanded.current = true;
+        terminal.resize(DEFAULT_TERMINAL_SIZE);
+      } else {
+        terminal.expand();
+      }
     } else {
-      terminal.expand();
+      terminalPanelRef.current?.collapse();
```

Comment:
Oh the inlining is nice. Those `showTerminal` and `hideTerminal` functions were not passed as props anywhere. This is simpler. I like it!